runtime/storage/firebase-storage.js

```diff
@@ -143,149 +167,557 @@ class FirebaseStorageProvider extends StorageProviderBase {
 
+// Models a Variable that is persisted to firebase in a
+// last-writer-wins scheme.
+//
+// Initialization: After construct/connect the variable is
+// not fully initialized, calls to `get` and `toLiteral`
+// will not complete until either:
+//  * The initial value is supplied via the firebase `.on`
+//    subscription.
+//  * A value is written to the store by a call to `set`.
+//
+// Updates from firebase: Each time an update is received
+// from firebase we update the local version and value,
+// unless there is a pending local modification (see below).
+//
+// Local modifications: When a local modification is applied
+// by a call to `set` we increment the version number,
+// mark this variable as locally modified, and start a
+// process to atomically persist the change to firebase.
+// Until this process has completed we suppress incoming
+// changes from firebase. The version that we have chosen
+// (by incrementing) may not match the final state that is
+// written to firebase (if there are concurrent changes in
+// firebase, or if we have queued up multiple local
+// modiciations), but the result will always be
+// monotonically increasing.
 class FirebaseVariable extends FirebaseStorageProvider {
   constructor(type, arcId, id, reference, firebaseKey) {
     super(type, arcId, id, reference, firebaseKey);
-    this.dataSnapshot = undefined;
-    this._pendingGets = [];
-    this._version = 0;
-    this.reference.on('value', dataSnapshot => {
-      this.dataSnapshot = dataSnapshot;
-      let data = dataSnapshot.val();
-      this._pendingGets.forEach(_get => _get(data));
-      this._pendingGets = [];
-      this._version = data.version;
-      this._fire('change', {data: data.data, version: data.version});
+
+    // Current value stored in this variable. Reflects either a
+    // value that was stored in firebase, or a value that was
+    // written locally.
+    this._value = null;
+
+    // Monotonic vesion, initialized via response from firebase,
+    // or a call to `set` (as 0). Updated on changes from firebase
+    // or during local modifications.
+    this._version = null;
+
+    // Whether `this._value` is affected by a local modification.
+    // When this is true we are still in the process of writing
+    // the value to the remote store and will suppress any
+    // incoming changes from firebase.
+    this._localModified = false;
+    
+    // Resolved when data is first available. The earlier of
+    // * the initial value is supplied via firebase `reference.on`
+    // * a value is written to the variable by a call to `set`.
+    this._initialized = new Promise(resolve => {
+      this._resolveInitialized = resolve;
     });
+
+    this.reference.on('value', dataSnapshot => this._remoteValueChanged(dataSnapshot));
+  }
+
+  _remoteValueChanged(dataSnapshot) {
+    if (this._localModified) {
+      return;
+    }
+    let data = dataSnapshot.val();
+    assert(this._version == null || data.version > this._version);
```

Comment:
When was the last time you saw that? I think that was a signal of the provider callbacks being re-entrant.

But here I think yes, this is safe: we're writing transactionally to firebase, and on each write we increase the version. We then rely on the `reference.on('value', ...` callback to provide us with changes in the same monotonic sequence. Since we never call `.off`, or otherwise re-establish the listener I think it's safe to assume it does.

Reply:
It was a while ago (not sure exactly how long). I poked at the current build and haven't seen it, so I guess it's gone.